### ISSUE TITLE
feat: implement tabular Q-learning for FrozenLake

### ISSUE THREAD

**Issue by ruddyscent** [ruddyscent/gymnasium-playbook#2]
## Goal
Implement a reproducible tabular Q-learning baseline for the deterministic 4×4 FrozenLake environment, the first step in the proposed learning path.

## Scope
- Use `FrozenLake-v1` with `map_name="4x4"` and `is_slippery=False`.
- Implement a Q-table, epsilon-greedy exploration, and the Bellman update without a reinforcement learning algorithm library.
- Separate training from evaluation and allow seeds and training hyperparameters to be configured.
- Handle episode termination and time-limit truncation correctly when computing learning targets.
- Document the environment, algorithm, dependencies, runnable commands, and observed results.

## Acceptance criteria
- [x] Training and evaluation can be reproduced using documented commands and seeds.
- [x] Evaluation uses a greedy policy without training updates and reports success rate and episode length.
- [x] Evaluation results are recorded for at least three training seeds and compared with a random-policy baseline.
- [x] Focused tests cover Q-learning targets, including terminal and truncated transitions.
- [x] Learned artifacts are saved under ignored output paths.

The learning path is a suggested educational sequence, not a delivery schedule or a list of completed implementations. This issue covers only the first environment.

## Acceptance verification

Implemented on [`codex/issue-2-frozenlake`](https://github.com/ruddyscent/gymnasium-playbook/tree/codex/issue-2-frozenlake) and merged in PR #4 as `35e65dbb3da53574df4fe4ae85a230a9b5aaf3fa`.

- **Reproducibility:** the [implementation README](https://github.com/ruddyscent/gymnasium-playbook/blob/35e65dbb3da53574df4fe4ae85a230a9b5aaf3fa/environments/toy_text/frozen_lake/README.md) documents locked uv setup, training, evaluation, seeds, and the three-seed benchmark. Repeatability tests pass.
- **Evaluation:** a separate greedy evaluator reports success rate and episode length; the regression test verifies that evaluation does not modify the Q-table.
- **Three seeds and baseline:** the [recorded results](https://github.com/ruddyscent/gymnasium-playbook/issues/2#issuecomment-5564941383) cover training seeds 0, 1, and 2. Each learned policy achieved 100% success with six mean steps; random policies achieved 1.5%, 1.2%, and 1.1%.
- **Learning targets:** all 14 regression tests passed again on September 7, 2026, including real terminal, time-limit truncation, and simultaneous termination/truncation transitions.
- **Artifacts:** default Q-tables, training histories, configuration, and benchmark reports are stored under `runs/frozen_lake/`; `git check-ignore` confirms the output path is ignored.

[PR CI](https://github.com/ruddyscent/gymnasium-playbook/actions/runs/34082553918) passed on Ubuntu 24.04, Windows 2022, and macOS 14 (Apple Silicon). This supersedes the earlier comment's pending-CI status. The full three-seed benchmark results were measured on macOS ARM64; CI checks environment smoke tests and regression tests, not the full benchmark or graphical replay on every platform.

**Comment by ruddyscent** [comment]
## Implementation insights

The local implementation on `codex/issue-2-frozenlake` produced the following observations. These are results for deterministic 4×4 FrozenLake, not claims about slippery environments or unseen maps. The branch has not yet been pushed or merged.

### Learning semantics

- **Termination and truncation need different Bellman targets.** Both end a rollout, but only `terminated` disables bootstrapping. A time-limit truncation still uses the next-state value. If the goal is reached on the last allowed step, both flags can be true; termination takes precedence. Tests exercise these cases using actual environment transitions.
- **Tie-breaking matters before any reward is discovered.** Every Q-value starts at zero. Always using `argmax` during exploitation favors action 0 (LEFT). Training instead samples uniformly among tied best actions. Evaluation uses a fixed greedy tie rule for reproducibility.
- **Training success and greedy evaluation success measure different behavior.** Epsilon remains at least 0.05, so training continues to make exploratory mistakes even after a successful route has been learned. The evaluator uses a fresh environment and never updates the Q-table.

### Measured results

Configuration: Python 3.13.15, Gymnasium 1.3.0, NumPy 2.5.3, macOS ARM64; `FrozenLake-v1`, `map_name="4x4"`, `is_slippery=False`, default rewards, 100-step limit. Each training run used 10,000 episodes, learning rate 0.1, discount 0.99, and epsilon `max(0.05, 1.0 * 0.999**episode)` with zero-based episode indexing. Each policy was evaluated for 1,000 episodes with seed `10000 + training_seed`.

| Training seed | First successful training episode | Successes in final 100 training episodes | Greedy evaluation success | Greedy mean steps | Random evaluation success | Random mean steps |
| --- | --- | --- | --- | --- | --- | --- |
| 0 | 38 | 91 | 100% | 6.000 | 1.5% | 7.556 |
| 1 | 31 | 97 | 100% | 6.000 | 1.2% | 7.857 |
| 2 | 30 | 93 | 100% | 6.000 | 1.1% | 7.642 |

Random-policy runs use the corresponding evaluation seed, not a trained table. Mean steps includes failed episodes: a short episode may simply mean falling into a hole.

For seed 0, the greedy route was `0 → 4 → 8 → 9 → 13 → 14 → 15`. Its six steps meet the Manhattan-distance lower bound. The learned start-state value for DOWN was approximately `0.95099005 = 0.99**5`, providing an interpretable check on discounting. The training code does not inspect the transition model or use this route as input.

Because the map, start state, and transitions are fixed, a greedy policy repeats the same trajectory. Repeating evaluation 1,000 times does not demonstrate generalization. The three training seeds show robustness to those exploration seeds only.

### Reproducibility and teaching

- Keep learning calculations separate from CLI/file handling, and preserve per-episode CSV records plus configuration/version metadata for later analysis and blog figures.
- After type annotations were added, all three seeds produced identical Q-tables, training CSV files, and evaluation results compared with the original implementation.
- Fourteen regression tests passed locally. Static checking passed for the seven Python files present at the type-annotation stage; the subsequently added viewer was not included in that static-check result. Its initialization/close path and CLI were checked, and the existing regression suite still passes.
- A `watch` command now replays the fixed policy or random play at an adjustable speed. This helps connect the table to observed behavior without mixing visualization into training.
- Ubuntu and Windows runtime verification remains pending CI. Local success and cross-platform dependency resolution are not substitutes for those runs.

Reproduce the comparison from the implementation worktree:

```sh
uv run --locked python -m environments.toy_text.frozen_lake benchmark --seeds 0 1 2 --episodes 10000 --eval-episodes 1000 --eval-seed 10000
```

The detailed explanation is maintained separately in `posts/frozen-lake-q-learning.md` in the local `blog` worktree. A future post can trace one update, explain the two ending flags, and compare exploratory training with greedy evaluation before discussing the fixed-map limitation.

**Comment by ruddyscent** [comment]
Closing after revalidating all five acceptance criteria against the merged implementation and current main. PR #4 is merged; its Ubuntu, Windows, and Apple Silicon macOS CI passed, and current main 0be6642ed3bd77b170b76f9d3f91ecf2836a42bd again passed the full three-platform FrozenLake semantics, type, and environment checks: https://github.com/ruddyscent/gymnasium-playbook/actions/runs/34578003435. The documented three-seed benchmark was also reproduced with the recorded learned-policy and random-baseline results. The full benchmark is not run on every CI platform and interactive policy playback is not covered by headless CI, but neither is an acceptance blocker.
